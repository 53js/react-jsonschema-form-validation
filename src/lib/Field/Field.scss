.Sjf_Field {
	&.isInvalid {
		&.isSubmitted,
		&.isTouched {
			border-color: $color-red;
		}
	}
}
