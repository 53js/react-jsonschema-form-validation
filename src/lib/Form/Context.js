import React, { useContext } from 'react';

const FormContext = React.createContext();

export default FormContext;

export const useFormContext = () => useContext(FormContext);

export const withFormContext = (cb) => <FormContext.Consumer>{cb}</FormContext.Consumer>;
