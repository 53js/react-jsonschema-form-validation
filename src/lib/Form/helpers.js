import Ajv from 'ajv';

/**
 * Returns a default Ajv instance
 */
export const createAjv = () => new Ajv({
	allErrors: true,
	v5: true,
	$data: true,
});

/**
 * Transforms an empty string or null to undefined
 * to allow usage of 'required' attribute in json schema.
 * Values undefined, null, or '' are equivalent in a form input.
 * @param {string} value
 */
export const empty = value => (
	(value === '' || value === null) ? undefined : value
);

/**
 * Formats data for Ajv validation
 * @param  {Object} data data to format
 * @return {Object}      formatted data
 */
export const formatData = (data) => {
	if (data instanceof Array) {
		return data.map(formatData);
	}

	if (data === Object(data)) {
		const copy = {}; // Do not modify original object !

		// eslint-disable-next-line no-restricted-syntax,guard-for-in
		for (const key in data) {
			// Usage of for-in to validate inherited properties
			// Example : File.name
			copy[key] = formatData(data[key]);
		}

		data = copy;
	}

	// Allow usage of Ajv's 'required' keyword
	data = empty(data);

	return data;
};

/**
 * Add a property field to the error object
 * The field value can be used to find errors on the object
 * based on a normalized path
 * @param {Array} errors Array of Ajv's errors
 */
export const formatErrors = errors => (errors || []).map((error) => {
	error.field = error.dataPath;

	if (error.keyword === 'required') {
		error.field = `${error.field}.${error.params.missingProperty}`;
	}

	error.field = error.field
		.replace(/^\./, '')
		.replace(/\[([0-9]+)\]/, '.$1');

	return error;
});

/**
 * Filters an array of string matching the fieldName expression
 * fieldName expression can be :
 * - the exact name of a field
 * - or a prefix if fieldName expression ends with '*'
 * @param {[string]} fields
 * @param {string} fieldName
 */
export const filterByFieldNameWithWildcard = (fields, fieldName) => {
	let regex;
	if (/\*$/.test(fieldName)) {
		regex = new RegExp(`^${fieldName.replace(/\*$/, '')}`);
	}

	return fields.filter((e) => {
		if (regex) {
			return regex.test(e.field);
		}
		return e.field === fieldName;
	});
};
