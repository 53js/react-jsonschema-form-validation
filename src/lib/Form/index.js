export { default } from './Form';
export { default as Context } from './Context';
