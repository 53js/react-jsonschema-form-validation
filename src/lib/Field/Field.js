import classnames from 'classnames';
import memoize from 'memoize-one';
import PropTypes from 'prop-types';
import React, { PureComponent } from 'react';

import Context from '../Form/Context';

class Field extends PureComponent {
	getOnBlurHandler = memoize((touch, name, onBlur) => (event) => {
		touch(name);
		if (onBlur) onBlur(event);
	});

	render() {
		const {
			children,
			className,
			component: Component,
			onBlur,
			name,
			forwardedRef,
			...props
		} = this.props;

		return (
			<Context.Consumer>
				{form => (
					<Component
						{...props}
						className={
							classnames(
								'Sjf_Field',
								className,
								{
									isInvalid: form.isFieldInvalid(name),
									isSubmitted: form.isSubmitted,
									isTouched: form.isFieldTouched(name),
								},
							)
						}
						name={name}
						onBlur={this.getOnBlurHandler(form.touch, name, onBlur)}
						ref={forwardedRef}
					>
						{ children }
					</Component>
				)}
			</Context.Consumer>
		);
	}
}

Field.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
	component: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
	forwardedRef: PropTypes.oneOfType([
		PropTypes.shape({}),
		PropTypes.func,
	]),
	onBlur: PropTypes.func,
	name: PropTypes.string.isRequired,
};

Field.defaultProps = {
	children: null,
	className: '',
	component: 'input',
	forwardedRef: null,
	onBlur: null,
};

// eslint-disable-next-line react/no-multi-comp
export default React.forwardRef((props, ref) => (
	<Field {...props} forwardedRef={ref} />
));
