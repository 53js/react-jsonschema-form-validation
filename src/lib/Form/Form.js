import Ajv from 'ajv';
import classnames from 'classnames';
import throttle from 'lodash.throttle';
import memoize from 'memoize-one';
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import scrollToElement from 'scroll-to-element';

import Context from './Context';
import {
	createAjv,
	filterByFieldNameWithWildcard,
	formatData,
	formatErrors,
	updateDataFromEvents,
} from './helpers';

class Form extends PureComponent {
	constructor(props) {
		super(props);

		const defaultAjv = props.ajv || createAjv();

		this.getValidator = memoize((ajv, schema, throttleDuration) => {
			ajv = ajv || defaultAjv;
			const validate = ajv.compile(schema);

			const validator = (data) => {
				const formattedData = formatData(data);
				const valid = validate(formattedData);
				const errors = formatErrors(validate.errors);

				this.setState({
					valid,
					errors,
				});
			};

			if (this.throttledValidator) this.throttledValidator.cancel();
			this.throttledValidator = throttle(validator, throttleDuration);

			return memoize(this.throttledValidator);
		});

		this.handleSubmit = this.handleSubmit.bind(this);

		/* eslint-disable react/no-unused-state */
		this.state = {
			errors: [],
			errorMessages: props.errorMessages,
			getFieldErrors: this.getFieldErrors.bind(this),
			handleFieldChange: this.handleFieldChange.bind(this),
			isFieldTouched: this.isFieldTouched.bind(this),
			isFieldInvalid: this.isFieldInvalid.bind(this),
			isInvalid: this.isInvalid.bind(this),
			isSubmitted: false,
			isTouched: this.isTouched.bind(this),
			touch: this.touch.bind(this),
			touchedFields: [],
			valid: true,
		};
		/* eslint-enable react/no-unused-state */
	}

	componentDidMount() {
		const { ajv, data, schema } = this.props;
		const validate = this.getValidator(ajv, schema);
		validate(data);
	}

	componentDidUpdate() {
		const { ajv, data, schema } = this.props;
		const validate = this.getValidator(ajv, schema);
		validate(data);
	}

	componentWillUnmount() {
		if (this.throttledValidator) this.throttledValidator.cancel();
	}

	getFieldErrors(fieldNames) {
		fieldNames = Array.isArray(fieldNames) ? fieldNames : [fieldNames];
		const { errors } = this.state;

		return fieldNames.reduce((fieldsErrors, fieldName) => [
			...fieldsErrors,
			...filterByFieldNameWithWildcard(errors, fieldName),
		], []);
	}

	handleFieldChange(event, value) {
		const { data, onChange } = this.props;
		if (onChange) {
			if (typeof event === 'string') {
				event = {
					target: {
						name: event,
						value,
					},
				};
			}
			const newData = updateDataFromEvents(data, event);
			onChange(newData, event);
		}
	}

	handleSubmit(event) {
		event.preventDefault();
		this.submit(event);
	}

	scrollToFirstError() {
		const { scrollOptions } = this.props;
		const { errors } = this.state;
		const firstError = errors[0];
		const element = document.getElementsByName(firstError.field)[0];

		scrollToElement(element, scrollOptions);
	}

	submit(event) {
		const {
			onSubmit,
			scrollToError,
		} = this.props;

		const {
			errors,
			valid,
		} = this.state;

		this.setState({ isSubmitted: true });

		if (valid) {
			this.reset();
			onSubmit(event);
		} else if (errors.length) {
			if (process.env.REACT_APP_JFV_DEBUG === 'true') console.log(errors); // eslint-disable-line no-console
			if (scrollToError) {
				this.scrollToFirstError();
			}
		}
	}

	isFieldTouched(fieldNames) {
		fieldNames = Array.isArray(fieldNames) ? fieldNames : [fieldNames];
		const { touchedFields } = this.state;
		return !!fieldNames.find(fieldName => (
			filterByFieldNameWithWildcard(touchedFields.map(field => ({ field })), fieldName).length > 0
		));
	}

	isTouched() {
		const { touchedFields } = this.state;
		return !!touchedFields.length;
	}

	isFieldInvalid(fieldNames) {
		fieldNames = Array.isArray(fieldNames) ? fieldNames : [fieldNames];
		return this.getFieldErrors(...fieldNames).length > 0;
	}

	isInvalid() {
		const { errors } = this.state;
		return !!errors.length;
	}

	reset() {
		this.setState({
			isSubmitted: false,
			touchedFields: [],
		});
	}

	touch(fieldNames) {
		fieldNames = Array.isArray(fieldNames) ? fieldNames : [fieldNames];
		const { touchedFields } = this.state;
		this.setState({
			touchedFields: [
				...new Set([
					...touchedFields,
					...fieldNames,
				]),
			],
		});
	}

	render() {
		const {
			ajv,
			children,
			className,
			component: FormComponent,
			data,
			throttleDuration,
			errorMessages,
			onChange,
			onSubmit,
			schema,
			scrollOptions,
			scrollToError,
			...rest
		} = this.props;

		const { isSubmitted } = this.state;

		const classes = classnames(
			className,
			{ isSubmitted },
		);

		return (
			<Context.Provider value={this.state}>
				<FormComponent
					className={classes}
					onSubmit={this.handleSubmit}
					noValidate
					{...rest}
				>
					{ children }
				</FormComponent>
			</Context.Provider>
		);
	}
}

Form.propTypes = {
	ajv: PropTypes.instanceOf(Ajv),
	children: PropTypes.node,
	className: PropTypes.string,
	component: PropTypes.elementType,
	data: PropTypes.shape({}),
	throttleDuration: PropTypes.number,
	errorMessages: PropTypes.shape({}),
	onChange: PropTypes.func,
	onSubmit: PropTypes.func.isRequired,
	schema: PropTypes.shape({}).isRequired,
	scrollToError: PropTypes.bool,
	scrollOptions: PropTypes.shape({}),
};

Form.defaultProps = {
	ajv: null,
	children: null,
	className: '',
	component: 'form',
	data: {},
	errorMessages: {},
	onChange: null,
	scrollToError: true,
	scrollOptions: {
		offset: 0,
		align: 'middle',
		duration: 900,
	},
	throttleDuration: 200,
};

export default Form;
