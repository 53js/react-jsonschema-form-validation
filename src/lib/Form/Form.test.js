import React from 'react';
import { shallow, mount } from 'enzyme';

import Form from './Form';
import Field from '../Field';

const testSchema = {
	type: 'object',
	properties: {
		type: { type: 'string', enum: ['te', 'ta'] },
	},
	required: [
		'type',
	],
};


it('should match snapshot', () => {
	const wrapper = shallow(<Form onSubmit={() => { }} schema={{}} />);
	expect(wrapper).toMatchSnapshot();
});

it('should call onSubmit handler when form submitted', () => {
	const onSubmit = jest.fn();
	const data = { type: 'te' };

	const wrapper = shallow(
		<Form
			data={data}
			onSubmit={onSubmit}
			schema={testSchema}
		/>,
	);

	wrapper.find('form').simulate('submit', { preventDefault() {} });
	expect(onSubmit).toHaveBeenCalled();
});

it('should not call onSubmit handler as the form is not valid', () => {
	const onSubmit = jest.fn();
	const data = { type: 'pioiv' };

	const wrapper = shallow(
		<Form
			data={data}
			onSubmit={onSubmit}
			schema={testSchema}
		/>,
	);

	wrapper.find('form').simulate('submit', { preventDefault() {} });
	expect(onSubmit).not.toHaveBeenCalled();
});

describe('Form.getFieldErrors()', () => {
	it('should return a list of fields having errors', () => {
		let data = { type: 'uuu' };

		let wrapper = shallow(
			<Form
				data={data}
				onSubmit={() => {}}
				schema={testSchema}
			/>,
		);

		expect(wrapper.instance().getFieldErrors('type').length).toStrictEqual(1);

		data = { type: 'te' };
		wrapper = shallow(
			<Form
				data={data}
				onSubmit={() => {}}
				schema={testSchema}
			/>,
		);
		expect(wrapper.instance().getFieldErrors('type').length).toStrictEqual(0);

		data = { type: null };
		wrapper = shallow(
			<Form
				data={data}
				onSubmit={() => {}}
				schema={testSchema}
			/>,
		);
		expect(wrapper.instance().getFieldErrors('type').length).toStrictEqual(1);
	});
});

describe('Form.isFieldTouched(fieldNames)', () => {
	it('should return true if at least one of the fields named on the list is touched, false otherwise', () => {
		const wrapper = mount(
			<Form
				onSubmit={() => {}}
				schema={{}}
			>
				<Field
					id="test-type"
					name="type"
					type="text"
				/>
				<Field
					id="test-name"
					name="name"
					type="text"
				/>
			</Form>,
		);

		expect(wrapper.instance().isFieldTouched(['type'])).toStrictEqual(false);
		expect(wrapper.instance().isFieldTouched(['name'])).toStrictEqual(false);
		expect(wrapper.instance().isFieldTouched(['type', 'name'])).toStrictEqual(false);

		wrapper.find('#test-type').hostNodes().simulate('blur', { preventDefault() {} });

		expect(wrapper.instance().isFieldTouched(['type'])).toStrictEqual(true);
		expect(wrapper.instance().isFieldTouched(['name'])).toStrictEqual(false);
		expect(wrapper.instance().isFieldTouched(['type', 'name'])).toStrictEqual(true);

		wrapper.find('#test-name').hostNodes().simulate('blur', { preventDefault() {} });

		expect(wrapper.instance().isFieldTouched(['type'])).toStrictEqual(true);
		expect(wrapper.instance().isFieldTouched(['name'])).toStrictEqual(true);
		expect(wrapper.instance().isFieldTouched(['type', 'name'])).toStrictEqual(true);
	});
});

describe('Form.isFieldInvalid(fieldNames)', () => {
	it('should return true if at least one of the fields named on the list has an error, false otherwise', () => {
		const data = {
			type: 'deon',
			name: 'Testing',
		};

		const newSchema = {
			type: 'object',
			properties: {
				type: { type: 'string', enum: ['te', 'ta'] },
				name: { type: 'string', minLength: 6 },
			},
			required: [
				'type',
				'name',
			],
		};

		let wrapper = shallow(
			<Form
				data={data}
				onSubmit={() => {}}
				schema={newSchema}
			/>,
		);

		expect(wrapper.instance().isFieldInvalid('type')).toBe(true);
		expect(wrapper.instance().isFieldInvalid(['name'])).toBe(false);

		const newData = {
			type: 'te',
			name: 'HE',
		};

		wrapper = shallow(
			<Form
				data={newData}
				onSubmit={() => {}}
				schema={newSchema}
			/>,
		);

		expect(wrapper.instance().isFieldInvalid('type')).toBe(false);
		expect(wrapper.instance().isFieldInvalid(['name'])).toBe(true);
	});
});

describe('Form.isInvalid()', () => {
	it('should return true if there is an error on the form, false otherwise', () => {
		const data = {
			type: 'sbbzrg',
			name: 'ee',
		};

		const newSchema = {
			type: 'object',
			properties: {
				type: { type: 'string', enum: ['te', 'ta'] },
				name: { type: 'string', minLength: 6 },
			},
			required: [
				'type',
				'name',
			],
		};

		const wrapper = shallow(
			<Form
				data={data}
				onSubmit={() => {}}
				schema={newSchema}
			/>,
		);

		expect(wrapper.instance().isInvalid()).toBe(true);
	});
});

describe('Form.isFieldTouched(fieldName)', () => {
	it('should return true if the field of name "fieldName" is touched, false otherwise', () => {
		const wrapper = mount(
			<Form
				onSubmit={() => {}}
				schema={{}}
			>
				<Field
					id="test-type"
					name="type"
					type="text"
				/>
			</Form>,
		);

		expect(wrapper.instance().isFieldTouched('type')).toStrictEqual(false);
		wrapper.find('#test-type').hostNodes().simulate('blur');
		expect(wrapper.instance().isFieldTouched('type')).toStrictEqual(true);
	});
});

describe('Form.isTouched()', () => {
	it('should return true if one of the fields in the form is touched, false otherwise', () => {
		const wrapper = mount(
			<Form
				onSubmit={() => {}}
				schema={{}}
			>
				<Field
					id="test-type"
					name="type"
					type="text"
				/>
				<Field
					id="test-name"
					name="name"
					type="text"
				/>
				<Field
					id="test-description"
					name="description"
					type="text"
				/>
			</Form>,
		);

		expect(wrapper.instance().isTouched()).toStrictEqual(false);
		wrapper.find('#test-type').hostNodes().simulate('blur', { preventDefault() {} });
		expect(wrapper.instance().isTouched()).toStrictEqual(true);
		wrapper.find('#test-name').hostNodes().simulate('blur', { preventDefault() {} });
		wrapper.find('#test-description').hostNodes().simulate('blur');
		expect(wrapper.instance().isTouched()).toStrictEqual(true);
	});
});

describe('Form.touch(fieldName)', () => {
	it('should add the field named "fieldName" with true value in the touched list in form state', () => {
		const wrapper = mount(
			<Form
				onSubmit={() => {}}
				schema={{}}
			>
				<Field
					id="test-type"
					name="type"
					type="text"
				/>
				<Field
					id="test-name"
					name="name"
					type="text"
				/>
				<Field
					id="test-description"
					name="description"
					type="text"
				/>
			</Form>,
		);

		expect(wrapper.state().touchedFields).toEqual([]);
		wrapper.instance().touch('type');
		expect(wrapper.state().touchedFields).toEqual(['type']);
		wrapper.instance().touch('name');
		expect(wrapper.state().touchedFields).toEqual(['type', 'name']);
		wrapper.instance().touch('description');
		expect(wrapper.state().touchedFields).toEqual(['type', 'name', 'description']);
	});
});
