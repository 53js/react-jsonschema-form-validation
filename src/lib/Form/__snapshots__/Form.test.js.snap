// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "errorMessages": Object {},
      "errors": Array [],
      "getFieldErrors": [Function],
      "isFieldInvalid": [Function],
      "isFieldTouched": [Function],
      "isInvalid": [Function],
      "isSubmitted": false,
      "isTouched": [Function],
      "touch": [Function],
      "touchedFields": Array [],
      "valid": true,
    }
  }
>
  <form
    className=""
    noValidate={true}
    onSubmit={[Function]}
  />
</ContextProvider>
`;
